File: leetcode/0734-sentence-similarity/solution.cpp
#include <vector>
#include <string>
#include <unordered_set>
using namespace std;

class Solution {
public:
    bool areSentencesSimilar(vector<string>& sentence1, vector<string>& sentence2, vector<vector<string>>& similarPairs) {
        if (sentence1.size() != sentence2.size()) {
            return false;
        }

        unordered_set<string> similar;
        for (auto& pair : similarPairs) {
            similar.insert(pair[0] + "#" + pair[1]);
            similar.insert(pair[1] + "#" + pair[0]);
        }

        for (size_t i = 0; i < sentence1.size(); i++) {
            const string& w1 = sentence1[i];
            const string& w2 = sentence2[i];
            if (w1 == w2) {
                continue;
            }
            if (similar.find(w1 + "#" + w2) == similar.end()) {
                return false;
            }
        }

        return true;
    }
};

File: leetcode/0734-sentence-similarity/Solution.java
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Solution {
    public boolean areSentencesSimilar(String[] sentence1, String[] sentence2, List<List<String>> similarPairs) {
        if (sentence1.length != sentence2.length) {
            return false;
        }

        Set<String> similar = new HashSet<>();
        for (List<String> pair : similarPairs) {
            similar.add(pair.get(0) + "#" + pair.get(1));
            similar.add(pair.get(1) + "#" + pair.get(0));
        }

        for (int i = 0; i < sentence1.length; i++) {
            String w1 = sentence1[i];
            String w2 = sentence2[i];
            if (w1.equals(w2)) {
                continue;
            }
            if (!similar.contains(w1 + "#" + w2)) {
                return false;
            }
        }

        return true;
    }
}
